src/test/java/com/google/api/codegen/testdata/csharp_gapic_resourcenames_library.baseline

```diff
@@ -0,0 +1,685 @@
+    }
+
+    public sealed partial class BookNameOneof : IResourceName
+    {
+        public enum Type
+        {
+            BOOK_NAME = 1,
+            ARCHIVED_BOOK_NAME = 2,
+            DELETED_BOOK_NAME_FIXED = 3,
+        }
+
+        public static BookNameOneof Parse(string name)
+        {
+            BookNameOneof result;
+            if (TryParse(name, out result))
+            {
+                return result;
+            }
+            throw new ArgumentException("Invalid name", "name");
+        }
+
+        public static bool TryParse(string name, out BookNameOneof result)
+        {
+            BookName bookName;
+            if (BookName.TryParse(name, out bookName))
+            {
+                result = new BookNameOneof(Type.BOOK_NAME, bookName);
+                return true;
+            }
+            ArchivedBookName archivedBookName;
+            if (ArchivedBookName.TryParse(name, out archivedBookName))
+            {
+                result = new BookNameOneof(Type.ARCHIVED_BOOK_NAME, archivedBookName);
+                return true;
+            }
+            DeletedBookNameFixed deletedBookNameFixed;
+            if (DeletedBookNameFixed.TryParse(name, out deletedBookNameFixed))
+            {
+                result = new BookNameOneof(Type.DELETED_BOOK_NAME_FIXED, deletedBookNameFixed);
+                return true;
+            }
+            result = null;
+            return false;
+        }
+
+        public static BookNameOneof From(BookName bookName) => new BookNameOneof(Type.BOOK_NAME, bookName);
+
+        public static BookNameOneof From(ArchivedBookName archivedBookName) => new BookNameOneof(Type.ARCHIVED_BOOK_NAME, archivedBookName);
+
+        public static BookNameOneof From(DeletedBookNameFixed deletedBookNameFixed) => new BookNameOneof(Type.DELETED_BOOK_NAME_FIXED, deletedBookNameFixed);
+
+        public BookNameOneof(Type type, IResourceName value)
+        {
+            _type = type;
+            _value = GaxPreconditions.CheckNotNull(value, nameof(value));
+        }
+
+        private Type _type;
+        private IResourceName _value;
+
+        // Possibly move to base class (if we have one?)
+        private T CheckAndReturn<T>(Type type)
+        {
+            if (_type != type)
+            {
+                throw new InvalidOperationException($"Requested type {type}, but this one-of contains type {_type}");
+            }
+            return (T)_value;
+        }
+
+        public BookName BookName => CheckAndReturn<BookName>(Type.BOOK_NAME);
+
+        public ArchivedBookName ArchivedBookName => CheckAndReturn<ArchivedBookName>(Type.ARCHIVED_BOOK_NAME);
+
+        public DeletedBookNameFixed DeletedBookNameFixed => CheckAndReturn<DeletedBookNameFixed>(Type.DELETED_BOOK_NAME_FIXED);
+
+        public override string ToString() => _value.ToString();
+    }
+
+    public sealed partial class DeletedBookNameFixed : IResourceName
```

Comment:
Would a base type be useful here? It would avoid code duplication - not sure how important that is. (A separate helper class could do that too, I suspect.)